### PR TITLE
Sample profiling

### PR DESCRIPTION
**Pull request by adrosenbaum**
Implementation of sample profiling.

User must first load 'profile variants' to the profile_variant collection
    loqusdb profile <vcf_file>

To activate the profiling function when using the load subcommand give the --check-profiles flag and --profile-threshold option
    loqusdb load ... --check-profiles --profile-theshold <float 0-1>

This will check the genotype of all samples in a vcf-file (not sv), at the positions of the profile_variants in the 'profile_variant' collection. The profile is assembled as a string for each sample. The profiles are checked against all profiles of previous samples in the database. If the check passes, the case and its variants are loaded. The profile string is found in the 'case' collection for each individual.

todo in this PR:
    tests

### CODE REVIEW

**Comment by moonso** [comment]
Hi @adrosenbaum , nice work! I'll wait for the tests and start with the review then. It makes it easier to see what is going on.

File: loqusdb/build_models/profile_variant.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+from loqusdb.models import ProfileVariant
+from .variant import get_variant_id
+
+LOG = logging.getLogger(__name__)
+
+def get_maf(variant):
+    """
+        if ID CAF exists in INFO column, return the allele frequency for
+        the alt allele
```
Comment (moonso) [inline loqusdb/build_models/profile_variant.py]: Could you follow the `Args: Returns:` syntax? Also can you elaborate a little bit more. What type is returned?

File: loqusdb/build_models/profile_variant.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    """
+
```
Comment (moonso) [inline loqusdb/build_models/profile_variant.py:20]: To make it more flat:
```python
if not variant.INFO.get('CAF'):
    return None
maf_list = json.loads(variant.INFO.get('CAF'))
return maf_list[1]
```

File: loqusdb/commands/load.py [hunk 3]
```diff
@@ -14,6 +14,12 @@
 
 LOG = logging.getLogger(__name__)
 
+def validate_profile_threshold(ctx, param, value):
+    if not (0 <= value <= 1):
+        raise ValueError('threshold must be between 0-1')
+    else:
```
Comment (moonso) [inline loqusdb/commands/load.py]:
```suggestion

```

File: loqusdb/commands/load.py [hunk 4]
```diff
@@ -14,6 +14,12 @@
...
+        return value
```
Comment (moonso) [inline loqusdb/commands/load.py]:
```suggestion
    return value
```

File: loqusdb/commands/load_profile.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+from . import base_command
+
+LOG = logging.getLogger(__name__)
+
+@base_command.command('profile', short_help = "Loads variants to be used in profiling")
+@click.argument('vcf_file', type = click.Path(exists=True))
+@click.pass_context
+def load_profile(ctx, vcf_file):
+
```
Comment (moonso) [inline loqusdb/commands/load_profile.py:47]: Please add a doctring with some more information like:
```python
"""
Takes a vcf with variants choosen for profile

Profiling is used when...
....
...
"""
```

File: loqusdb/plugins/mongo/adapter.py [hunk 6]
```diff
@@ -4,14 +4,14 @@
 from loqusdb.plugins import Base
 from mongo_adapter import MongoAdapter as BaseAdapter
 
-from . import VariantMixin, CaseMixin
+from . import VariantMixin, CaseMixin, ProfileVariantMixin
```
Comment (moonso) [inline loqusdb/plugins/mongo/adapter.py]:
```suggestion
from . import (VariantMixin, CaseMixin, ProfileVariantMixin)
```

File: loqusdb/plugins/mongo/profile_variant.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import logging
+
+LOG = logging.getLogger(__name__)
+
+class ProfileVariantMixin():
+
+    def add_profile_variants(self, profile_variants):
+
+        """Add several variants to database
```
Comment (moonso) [inline loqusdb/plugins/mongo/profile_variant.py]:
```suggestion
        """Add several variants to the profile collection in the database
```

File: loqusdb/utils/profiling.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+
+    return profiles
+
+def profile_match(adapter, profiles, threshold=0.9):
+
+    """
+        given a dict of profiles, searches through all the samples in the DB
+        for a match.
+
```
Comment (moonso) [inline loqusdb/utils/profiling.py:84]: Could you write some more about what is happening when a match is found?

**Review by moonso — CHANGES_REQUESTED** [review]
Nice work @adrosenbaum !! I like the testing as well.
There where some comments, minor stuff. Could you add some stuff to the documentation? It needs to be explained somewhere why this is here and how it works

**Comment by adrosenbaum** [comment]
Fixed what you commented on @moonso. 

I've also added a functionality that will update the profiles of existing samples in the database, if a path to a snv VCF file is given for the case. (loqusdb profile --update).

There is also a feature 'loqusdb profile --stats' that I thought should print some statistics based on the profiles. I am however unsure of what statistics might be useful to look at (or if this feature is necessary at all). Now it calculates the pair wise hamming distance between all sample profiles in the DB, and calculates the mean, and std, to get an idea of how well the samples can be distinguished from each other using this method.

**Comment by adrosenbaum** [comment]
Pushed some new commits.

Do you have the possibility to review this soon @moonso? It would be nice to test this branch tomorrow on stage.

**Review by moonso — APPROVED** [review]

**Comment by adrosenbaum** [comment]
Hopefully final change.

When loading a case, the vcf that the profiling is made on is entered separate from the variants that is to be loaded into the db. The 'profile' vcf is given with the --check-profile option. Any thoughts @moonso and @henrikstranneheim?

**Comment by henrikstranneheim** [comment]
@moonso Do you have time to look at this PR today?

**Comment by moonso** [comment]
Ok I'll have a look sometime today!

**Comment by henrikstranneheim** [comment]
👍

**Review by moonso — APPROVED** [review]
Looks great @adrosenbaum ! All I would like now are a couple of lines in the documentation about the idea of the profiling feature. Why it is there and why one needs to point to another VCF than the load vcf. This might be clear to us but if another user is uploading it would be nice to have it in the docs. 👍

**Comment by moonso** [comment]
Oh and can you update the CHANGELOG.md?

**Comment by adrosenbaum** [comment]
Added to documentation and CHANGELOG.md. Does this sufficiently describe the new profiling feature @moonso? If so I'll merge with master.

**Comment by moonso** [comment]
Great! Go ahead

**Pull request merged**